README.md
redmine_contacts_helpdesk_gpg
=============================

Requirements
------------

* Redmine 3.4-4.0
* RedmineCRM's Contacts plugin (pro or light)
* RedmineCRM's Helpdesk plugin (pro)

This plugin does not provide a (licensed) copy of RedmineCRM's plugins. Please get your own. :)

Installation
------------

### Prepare OS

Install these packages

* gnupg_1.4.xx
* optional: gpgsm (on some OS bundled with gnupg)
* libgpgme11-dev (Name may vary depending on OS. It's the package providing executable 'gpgme-config')


### Install required GEMs

  gem install gpgme -- --use-system-libraries
  gem install mail-gpg

Please install gpgme with '--use-system-libraries' because otherwise gpgme would compile its own versions of gpg libraries.

Further development of this plugin expects gpgsm to be available and I hate mixing libraries from different sources.


### Install this plugin

Unarchive plugin to redmine/plugins

  rake redmine:plugins RAILS_ENV=production


Uninstallation
--------------

  rake redmine:plugins NAME=redmine_contacts_helpdesk_gpg VERSION=0 RAILS_ENV=production
  rm -rf plugins/redmine_contacts_helpdesk_gpg


Setup
-----

### Params

* Directory containing key rings - Default is determined by environment GNUPGHOME or fallback "~/.gnupg"
* URL of public key server - Only "http://pool.sks-keyservers.net:11371" has been tested yet. Other keyservers might work...


### Add cron jobs for keystore maintenance

Paths in the following examples should be set according to your environment.

Cron jobs should be run as the same user running redmine (e.g. www-data)

  # refresh keys from keyserver on the first day of the month
  00 00 01 * * cd /srv/redmine && bundle exec rake RAILS_ENV=production redmine:plugins:helpdesk_gpg:refresh_keys
  # delete expired keys from keystore on the 15th
  00 00 15 * * cd /srv/redmine && bundle exec rake RAILS_ENV=production redmine:plugins:helpdesk_gpg:remove_expired_keys


Credits
-------

### Icons

* Icons are taken from "FatCow-Farm Fresh Icons" (http://www.fatcow.com/free-icons) by FatCow Web Hosting (http://www.fatcow.com) and are licensed under CC BY US 3.0 (http://creativecommons.org/licenses/by/3.0/us/)
* Or were inspired by above icons

License
-------

This redmine plugin is released under the MIT license.


CHANGELOG.md
Changelog
=========

20.05.0
-------

* Add admin setting for path to GPG binaries.

20.03.1
-------

* Fix missing information if initial outgoing mail was encrypted/signed.
* Validate if encrypted keys are present for initial mails before saving the issue.

20.03.0
-------

* Supports Helpdesk 4.1.6 and Redmine 4.1.
* Fix mail not being sent because first recipient is empty.

19.12.0
-------

* Fix handling incoming signed mails, ignore verification errors.
  Signature verification errors are handled like an invalid signature.


19.11.1
-------

* Catch and log GPGME errors.
* Improve mail (in/out) logging to catch errors in the GPG and helpdesk plugins.
* Add some missing German translations.

* Fix error for new helpdesk issues when contact has been created in the issue.

19.11.0
-------

* Fix key status info for Helpdesk 4.1.5.

19.10.0
-------

* Remove support for GPG passphrases.


0.5.1
-----

* Fix saving of issues that were encrypted initially.

0.5.0 (Jun 2019)
----------------

* Fix for mail area customizations for Helpdesk 4.1.0.
* GPG options checkboxes are now on the same line as the mail checkbox above the mail area.
* Always show full recipient fields, CC / BCC addresses are always visible now.
* Fix display of GPG key indicators for the recipient fields.
* Hide GPG options if mail checkbox is not checked.
* Give earlier feedback about missing keys and don't save journal then.

0.4.0 (Jun 2019)
----------------

* Preselect encryption when replying to an encrypted issue.
* Fix gpg send options checkbox labels.
* code cleanups, modernization
* Fix key lookup from Javascript code.
* target Ruby 2.5

0.3.0 (Apr 2019)
----------------

* Redmine 4.x support
* Merge code from https://github.com/alexandermeindl/redmine_contacts_helpdesk_gpg
  * fixed key management

0.2.0 (Jan 2019)
----------------

* Helpdesk 4.0 support

0.1.0 (Nov 2018)
----------------

* Redmine 3.x support

0.0.7 (Oct 2018)
----------------

* fixed typo
* fixed routing for redmine v3.x
* added missing changelog entries
* checking keys in bcc

0.0.6 (Sep 2015)
----------------

* small fix

0.0.5 (Sep 2015)
----------------

* code refactoring

0.0.4 (May 2015)
----------------

* per project setting: sign/encrypt by default
* enable signing/encryption for new tickets

0.0.3 (March 2015)
----------------

*   Code cleanup
*   Initial checkin

0.0.2 unreleased
----------------

*   added rake tasks for key management

0.0.1 unreleased
----------------

*   initial development version



init.rb
require 'redmine_contacts_helpdesk_gpg'

Rails.logger.info 'Starting GPG Helper Plugin for RedmineUP\'s Helpdesk Plugin'

# Plugin definition
Redmine::Plugin.register :redmine_contacts_helpdesk_gpg do
  name 'Redmine Contacts Helpdesk GPG'
  author 'darkstarSH / Alphanodes GmbH / Tobias Stenzel'
  description 'This is a plugin for Redmine to use GPG signing/encryption in RedmineUP\'s helpdesk'
  version '20.05.0'
  url 'https://github.com/piratenpartei/redmine_contacts_helpdesk_gpg'

  requires_redmine version_or_higher: '4.0'
  requires_redmine_plugin :redmine_contacts, version_or_higher: '4.2.1'
  requires_redmine_plugin :redmine_contacts_helpdesk, version_or_higher: '4.1.0'

  settings default: {
    gpg_keyrings_dir: ENV['GNUPGHOME'] || '~/.gnupg',
    gpg_keyserver: 'http://pool.sks-keyservers.net:11371',
    gpg_bin_dir: ''
  }, partial: 'settings/gpg_settings'

  menu :admin_menu, :gpg_keystore, { controller: 'gpgkeys', action: 'index' }, caption: :label_gpg_keystore, param: nil, html: { class: 'icon' }
end

if ActiveRecord::Base.connection.table_exists?(:settings)
  Rails.configuration.to_prepare do
    HelpDeskGPG.setup
  end
end
